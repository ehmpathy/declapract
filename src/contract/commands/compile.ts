import { Command, flags } from '@oclif/command';

import { compile } from '../../logic/commands/compile';

export default class Compile extends Command {
  public static description =
    'compile the declared declarations so that they can be packaged and distributed by npm safely';

  public static flags = {
    help: flags.help({ char: 'h' }),
    sourceDirectory: flags.string({
      char: 's',
      description: 'the source directory which contains the declarations to compile',
      required: true,
      default: 'src',
    }),
    distributionDirectory: flags.string({
      char: 'd',
      description: 'the distribution directory to which we will compile the declarations',
      required: true,
      default: 'src',
    }),
  };

  public async run() {
    const { flags } = this.parse(Compile);
    const sourceDirectory = flags.sourceDirectory!;
    const distributionDirectory = flags.distributionDirectory;

    // generate the code
    const sourceDirectoryAbsolute =
      sourceDirectory.slice(0, 1) === '/' ? sourceDirectory : `${process.cwd()}/${sourceDirectory}`; // if starts with /, consider it as an absolute path
    const distributionDirectoryAbsolute =
      distributionDirectory.slice(0, 1) === '/' ? distributionDirectory : `${process.cwd()}/${distributionDirectory}`; // if starts with /, consider it as an absolute path
    await compile({
      sourceDirectory: sourceDirectoryAbsolute,
      distributionDirectory: distributionDirectoryAbsolute,
    });
  }
}
