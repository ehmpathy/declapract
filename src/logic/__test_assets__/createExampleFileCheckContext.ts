import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
