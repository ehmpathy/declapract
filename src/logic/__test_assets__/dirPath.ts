export const testAssetsDirectoryPath = __dirname;
