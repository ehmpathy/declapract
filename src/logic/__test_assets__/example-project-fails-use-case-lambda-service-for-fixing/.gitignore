# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

