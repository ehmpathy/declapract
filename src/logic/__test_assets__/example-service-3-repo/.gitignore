.declapract
node_modules
package-lock.json
