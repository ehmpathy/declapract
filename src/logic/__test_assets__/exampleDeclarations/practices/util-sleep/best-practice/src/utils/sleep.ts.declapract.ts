export const check = { optional: true };
