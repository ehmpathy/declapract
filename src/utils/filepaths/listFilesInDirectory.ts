import glob from 'glob';
import { promisify } from 'util';

const promisifiedGlob = promisify(glob);

/**
 * returns paths to all files inside this directory or its nested directories
 */
export const listFilesInDirectory = async ({ directory }: { directory: string }) =>
  promisifiedGlob('**/*', { cwd: directory, dot: true, nodir: true });
