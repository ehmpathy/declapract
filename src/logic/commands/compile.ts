import { copyFileAsync } from '../../utils/fileio/copyFileAsync';
import { listFilesInDirectory } from '../../utils/filepaths/listFilesInDirectory';

export const serializeGlobPathForNpmPackaging = (path: string) => path.replace(/\*/g, '<star>');
export const deserializeGlobPathFromNpmPackaging = (path: string) => path.replace(/\<star\>/g, '*');

/**
 *
 * compiles declarations from a `sourceDirectory` into `distributionDirectory`, making sure that they can be packaged and used with npm successfully
 *
 * specifically, while copying each file from `sourceDirectory` to `distributionDirectory`:
 * - replaces file path `*` chars to `<star>` chars (https://github.com/uladkasach/declapract/issues/5)
 *
 * note: this is expected to be temporarily required functionality
 * - hopefully, we will be able to remove this when npm supports `*` in path names itself
 * - https://github.com/npm/cli/issues/3722
 */
export const compile = async ({
  sourceDirectory,
  distributionDirectory,
}: {
  sourceDirectory: string;
  distributionDirectory: string;
}) => {
  // list all the files in the source dir
  const filePaths = await listFilesInDirectory({ directory: sourceDirectory });
  console.log(`📝 compiling ${filePaths.length} files...`);

  // write each one to the distribution directory
  await Promise.all(
    filePaths.map((filePath) =>
      copyFileAsync({
        from: `${sourceDirectory}/${filePath}`,
        to: `${distributionDirectory}/${serializeGlobPathForNpmPackaging(filePath)}`,
      }),
    ),
  );
};
