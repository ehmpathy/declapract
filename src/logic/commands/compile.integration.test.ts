import { removeDirectoryAsync } from '../../utils/fileio/removeDirectoryAsync';
import { listFilesInDirectory } from '../../utils/filepaths/listFilesInDirectory';
import { log } from '../../utils/logger';
import { testAssetsDirectoryPath } from '../__test_assets__/dirPath';
import { compile } from './compile';

const logSpy = jest.spyOn(console, 'log').mockImplementation(() => log.debug); // swap to log debug so its not displaying during tests by default

describe('compile', () => {
  it("should be able to correctly compile declarations, with '*' substitution", async () => {
    const srcDir = `${testAssetsDirectoryPath}/example-best-practices-compile-for-package-repo/src`;
    const distDir = `${testAssetsDirectoryPath}/example-best-practices-compile-for-package-repo/dist`;

    // clean up the test env
    await removeDirectoryAsync({
      directory: distDir,
    });

    // compile
    await compile({
      sourceDirectory: srcDir,
      distributionDirectory: distDir,
    });

    // and check that the files are correctly defined
    const filesInDistDir = await listFilesInDirectory({ directory: distDir });
    expect(filesInDistDir).toContain(
      'practices/directory-structure-src/best-practice/src/logic/<star><star>/<star>.ts.declapract.ts',
    );
    expect(filesInDistDir).toMatchSnapshot();

    // and check display output
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
});
