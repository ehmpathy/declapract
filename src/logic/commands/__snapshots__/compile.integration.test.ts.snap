// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile should be able to correctly compile declarations, with '*' substitution 1`] = `
[
  "examples/typescript/.gitignore",
  "examples/typescript/.prettierignore",
  "examples/typescript/declapract.use.yml",
  "examples/typescript/package.json",
  "examples/typescript/prettier.config.js",
  "examples/typescript/src/domain/index.ts",
  "getVariables.ts",
  "practices/directory-structure-src/bad-practices/model-dir/.declapract.readme.md",
  "practices/directory-structure-src/bad-practices/model-dir/src/model/<star><star>/<star>.ts.declapract.ts",
  "practices/directory-structure-src/bad-practices/services-dir/src/services/<star><star>/<star>.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/.declapract.readme.md",
  "practices/directory-structure-src/best-practice/src/contract/<star><star>/<star>.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/src/data/clients/<star><star>/<star>.ts",
  "practices/directory-structure-src/best-practice/src/data/clients/<star><star>/<star>.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/src/data/dao/<star><star>/<star>.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/src/domain/constants.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/src/domain/index.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/src/domain/objects/index.ts.declapract.ts",
  "practices/directory-structure-src/best-practice/src/logic/<star><star>/<star>.ts.declapract.ts",
  "practices/prettier/best-practice/.prettierignore",
  "practices/prettier/best-practice/package.json",
  "practices/prettier/best-practice/package.json.declapract.ts",
  "practices/prettier/best-practice/prettier.config.js",
  "useCases.yml",
]
`;

exports[`compile should be able to correctly compile declarations, with '*' substitution 2`] = `
[
  [
    "üìù compiling 24 files...",
  ],
]
`;
